src/core/namespace.c

```diff
@@ -3306,6 +3308,336 @@ bool ns_type_supported(NamespaceType type) {
         return access(ns_proc, F_OK) == 0;
 }
 
+static int get_unpeeled_mount_fd(const char *mount_path, int *ret_fd) {
+        _cleanup_close_pair_ int pipe_fds[2] = EBADF_PAIR;
+        pid_t pid;
+        int r;
+
+        assert(mount_path);
+        assert(ret_fd);
+
+        r = socketpair(AF_UNIX, SOCK_DGRAM|SOCK_CLOEXEC, 0, pipe_fds);
+        if (r < 0)
+                return log_debug_errno(errno, "Failed to create socket pair: %m");
+
+        r = safe_fork("(sd-ns-unpeel)", FORK_DEATHSIG_SIGTERM|FORK_LOG|FORK_WAIT, &pid);
+        if (r < 0)
+                return r;
+        if (r == 0) {
+                _cleanup_close_ int dir_fd = -EBADF;
+
+                pipe_fds[0] = safe_close(pipe_fds[0]);
+
+                /* Clone mount namespace here to unpeel without affecting live process */
+                r = detach_mount_namespace();
```

Comment:
What is that unpeeled fd used for exactly?

Afaict, you have a current overlayfs "ovl1" mounted on top of /etc and you want to replace this with a new overlayfs mount "ovl2".

The way I would expect this to work is as follows:

fd_ovl_context = fsopen(ovl)
// set up with all mount options an layers and so on
fsconfig(fd_ovl_context, FSCONFIG_CMD_CREATE)
fd_ovl_mnt = fsmount(fd_ovl_context)

umount2(mount_path)

move_mount(fd_ovl_mnt, mount_path)

What do yo uneed that unmount for via get_unpeeled_fd()?

Reply:
IOW, why do you care about mount_path aka the target mountpoint at all before you've fully assembled the new ovl2 mount?

Reply:
we need to find the base layer in order to use it in the new overlay

Reply:
> What is that unpeeled fd used for exactly?
>
> Afaict, you have a current overlayfs "ovl1" mounted on top of /etc and you want to replace this with a new overlayfs mount "ovl2".

Yeah, we do this less for the mount exchange part (which MOVE_MOUNT_BENEATH nicely exists for it) and more so for assembling ovl2 to start with. 

The layers look like `/extension1:/extension2:/extension3:/etc`. But if you refer to e.g. /etc directly when setting up layers, you would be layering over the old already overlayed /etc. We do the fake unmount so that we can get the original /etc and be able to refer to it by FD. This was the idea that got workshopped in [this thread](https://github.com/systemd/systemd/pull/33995#discussion_r1793485032).

Reply:
I see, /etc is part of your layer stack.